index: handle case where pindex_prev equals chain tip in NextSyncBlock()

HowHsu opened the pull request:
When `pindex_prev` is the chain tip, `NextSyncBlock()` should return `nullptr` (there's no next block to sync). Currently this works, but relies on the fork/reorg path to produce the result: `chain.Next(tip)` returns `nullptr`, falling through to `FindFork(tip)` which returns `tip` itself, then `chain.Next(tip)` returns `nullptr` again.
Add an explicit early return for this case. This makes `NextSyncBlock()`'s three cases self-documenting:

1. next block exists on the active chain — return it
2. at the chain tip — return `nullptr`
3. on a stale branch — find the fork point and return the block after it

It also makes the existing comment ("Since block is not in the chain") accurate — the tip is in the chain, so it shouldn't reach that path.

stickies-v reviewed (commented):
You're correct that the current docstring is incorrect. I personally don't find your suggestion very clear either, as the two cases talk about different parts of the return statement.

An alternative would be to keep the docstring as-is and just make the chaintip case more explicit in code:

```cpp
    const CBlockIndex* pindex = chain.Next(pindex_prev);
    if (pindex || pindex_prev == chain.Tip()) {
        return pindex;
    }
```

<details>
<summary>git diff on d6bf3b1cc2</summary>

```diff
diff --git a/src/index/base.cpp b/src/index/base.cpp
index 8c958ac5f6..506ee73d2e 100644
--- a/src/index/base.cpp
+++ b/src/index/base.cpp
@@ -140,15 +140,12 @@ static const CBlockIndex* NextSyncBlock(const CBlockIndex* pindex_prev, CChain&
     }
 
     const CBlockIndex* pindex = chain.Next(pindex_prev);
-    if (pindex) {
+    if (pindex || pindex_prev == chain.Tip()) {
         return pindex;
     }
 
-    // Two cases goes here:
-    // 1. pindex_prev is the tip: chain.FindFork(pindex_pre) returns pindex_prev itself
-    // 2. pindex_prev is not in m_chain: since block is not in the chain, return the next
-    // block in the chain AFTER the last common ancestor. Caller will be responsible for
-    // rewinding back to the common ancestor.
+    // If block is not in the chain, return the next block in the chain AFTER the last common ancestor.
+    // Caller will be responsible for rewinding back to the common ancestor.
     return chain.Next(chain.FindFork(pindex_prev));
 }
 

```
</details>

HowHsu committed:
index: add explicit early exit in NextSyncBlock() when the input is the chain tip

When pindex_prev is the chain tip, return earlier and explicitly rather than
mixing it with the reorg case. For more detail, please see PR:
https://github.com/bitcoin/bitcoin/pull/32875

Co-authored-by: l0rinc <pap.lorinc@gmail.com>
Co-authored-by: optout <optout@nostrplebs.com>
Co-authored-by: furszy <matiasfurszyfer@protonmail.com>

l0rinc commented on src/index/base.cpp:
```diff
@@ -140,7 +140,7 @@ static const CBlockIndex* NextSyncBlock(const CBlockIndex* pindex_prev, CChain&
 
     const CBlockIndex* pindex = chain.Next(pindex_prev);
-    if (pindex) {
+    if (pindex || pindex_prev == chain.Tip()) {
```
seems you've found a code that isn't covered by tests - could you add a test case which showcases the need for this addition (such that if we revert this code the test fails)?

l0rinc replied:
(if you delete and repost instead of editing, we're getting multiple emails. if we need to modify it, we just cross out old version and add `edit:` to add the new part)

l0rinc replied:
Not sure what that means, we write tests to tell us when the behavior of the application changed.
You have changed the behavior and no test broke, so there's a lack of test coverage that we should fix to explain why this change is needed. It also helps the reviewers debug the given scenario to understand why this changes is needed - or if there's a different solution that would also solve it, etc.

HowHsu replied:
> Not sure what that means, we write tests to tell us when the behavior of the application changed. You have changed the behavior and no test broke, so there's a lack of test coverage that we should fix to explain why this change is 

this PR doesn't change the behavior of `NextSyncBlock()`, it just tweaks the code to make the function clearer.

HowHsu replied:
> > If the current code is already covered by tests (I haven't checked), it should be explained in the PR description (which is already too vague, it's not obvious what problem we're solving). If not, we should add a dedicated test. Please unresolve this comment, I don't consider my concern addressed.
> 
> Sure, I0rinc, I'll update the description. As @optout21 said, it's a slight optimization, so it doesn't change the functionality of `NextSyncBlock()`, So I don't have clues how to test the change, but I think we can add tests for the whole method `NextSyncBlock()` since there is not any unit tests for this method now.

@l0rinc description of PR updated.

l0rinc commented on src/index/base.cpp:
```diff
@@ -155,8 +155,11 @@ static const CBlockIndex* NextSyncBlock(const CBlockIndex* pindex_prev, CChain&
+        return nullptr;
+    }
+
+    if (const CBlockIndex* pindex = chain.Next(pindex_prev); pindex) {
```
```suggestion
    if (auto* pindex{chain.Next(pindex_prev)}) {
```

l0rinc commented on src/index/base.cpp:
PR description doesn't explain what `chain.Next` or `CChain::Contains` does in case of `chain.Tip()` or how we can test this manually or automatically.
It could also mention https://github.com/bitcoin/bitcoin/blob/b58eebf1520f503bd4c7c5693546a6859064ce51/src/index/base.cpp#L220-L221 as an existing explanation.
There's also not problem description, where is this code used, what are we looking at (to help reviewers and to prove you understood the change), what we're trying to solve exactly (`CChain::FindFork` meant for reorgs, it's walking the ancestor chain which is more expensive for tip than it needs to be).

@optout21's framing is best: "a slight optimization -- early exit handling of a special case" 

It's also stating it "handles a specific case in a wrong way" which is too strong: the behavior was correct, just hacky - it still returned `nullptr`, it's just very hacky to do it via out of bounds `[pindex->nHeight + 1]`.

l0rinc reviewed (changes requested):
Concept ACK, we shouldn't use the hacky out of bounds method to return for a common scenario, adding an explicit tip check documents the code better.

But approach NACK, I'm not convinced the author understands the problem, so I have added additional info and tests that could be incorporated into the commits and PR description.

furszy reviewed (commented):
ACK 7d2a7597af2be8edec7473c8360849a7ab835cc9

This isn't really an optimization as the first commit suggests. The reorg and tip paths run rarely, and this adds a `Tip()` call for every block during initial sync (which translates to an extra vector lookup per block).

Still, this makes `NextSyncBlock()`'s behavior explicit, so merging it isn’t that bad either. We have #34489, which improves the situation by avoiding calls to `NextSyncBlock()` anyway.

furszy reviewed (commented):
ACK with a nuance.

HowHsu commented on src/index/base.cpp:
```diff
@@ -155,11 +155,15 @@ static const CBlockIndex* NextSyncBlock(const CBlockIndex* pindex_prev, CChain&
         return pindex;
     }
 
+    // if at the tip, there is no next
```
done

HowHsu commented on src/test/txindex_tests.cpp:
```diff
@@ -31,6 +31,10 @@ BOOST_FIXTURE_TEST_CASE(txindex_initial_sync, TestChain100Setup)
     txindex.Sync();
 
+    // BlockUntilSyncedToCurrentChain should return true after Sync() completes,
+    // since all historical blocks have been indexed.
```
I see, that commit has been removed. Thanks, furszy.

furszy reviewed (commented):
ACK db3c25cfae1cc70171cc9bd921a2bf9be2c9da7e

sedited commented:
@l0rinc can you respond to @HowHsu's latest comment here? If you still have concerns, maybe re-enumerate them again.

HowHsu commented:
> Nothing has changed, he's still [ignoring my comments](https://github.com/bitcoin/bitcoin/pull/32875#discussion_r2758871724), and claims tests were added, even though this PR doesn't have any. The author doesn't understand the basics; I don't think he should touch this code.

Please see this conversation: https://github.com/bitcoin/bitcoin/pull/32875#discussion_r2777101205
The first type of tests you mentioned are inclued by @optout21 's PR, so I removed my previous added one.
Please see this conversation: https://github.com/bitcoin/bitcoin/pull/32875#discussion_r2941200576
the second type of tests you mentioned to add in like txindex_tests was added, but just like what @furszy said, it's actually orthogonal to the goal of this PR. 
 Let me clarify: this PR is primarily a code cleanup to make the logic clearer, with a small optimization as well—the special case handled here now returns early.
How about you providing a test as you described `such that if we revert this code the test fails`

stickies-v reviewed (approved):
ACK db3c25cfae1cc70171cc9bd921a2bf9be2c9da7e

This is a pretty trivial refactor that makes this function much easier to understand. I frankly don't understand the pushback (on the current version)

---

@HowHsu I think it's great you're doing everything you can to make this PR easy-to-review. That said, I think you might have taken it a bit further than is productive: the current PR description makes this seem like way bigger of a change/deal than it is. Suggested alternative you're free to use, if you think it's good:


> When `pindex_prev` is the chain tip, `NextSyncBlock()` should return `nullptr` (there's no next block to sync). Currently this works, but relies on the fork/reorg path to produce the result: `chain.Next(tip)` returns `nullptr`, falling through to `FindFork(tip)` which returns `tip` itself, then `chain.Next(tip)` returns `nullptr` again.
> 
> Add an explicit early return for this case. This makes `NextSyncBlock()`'s three cases self-documenting:
>  1. next block exists on the active chain — return it
>  2. at the chain tip — return `nullptr`
>  3. on a stale branch — find the fork point and return the block after it
> 
> It also makes the existing comment ("Since block is not in the chain") accurate — the tip is in the chain, so it shouldn't reach that path.
